sdk/identity/azure-identity/azure/identity/_internal/aad_client_base.py

```diff
@@ -96,17 +89,43 @@ def _process_response(self, response, scopes, now):
             )
         return AccessToken(response_copy["access_token"], expires_on)
 
-    @abc.abstractmethod
-    def _get_client_session(self, **kwargs):
-        pass
-
-    @abc.abstractmethod
-    def _obtain_token(self, scopes, fn, **kwargs):
-        # type: (Iterable[str], Callable, **Any) -> AccessToken
-        pass
+    def _get_auth_code_request(self, scopes, code, redirect_uri, client_secret=None):
+        # type: (str, str, Sequence[str], Optional[str]) -> HttpRequest
+
+        data = {
+            "client_id": self._client_id,
+            "code": code,
+            "grant_type": "authorization_code",
+            "redirect_uri": redirect_uri,
+            "scope": " ".join(scopes),
+        }
+        if client_secret:
+            data["client_secret"] = client_secret
+
+        request = HttpRequest(
+            "POST", self._token_endpoint, headers={"Content-Type": "application/x-www-form-urlencoded"}
+        )
+        request.set_formdata_body(data)
+        return request
+
+    def _get_refresh_token_request(self, scopes, refresh_token):
+        # type: (str, Sequence[str]) -> HttpRequest
+
+        data = {
+            "grant_type": "refresh_token",
+            "refresh_token": refresh_token,
+            "scope": " ".join(scopes),
+            "client_id": self._client_id,
+        }
+        request = HttpRequest(
+            "POST", self._token_endpoint, headers={"Content-Type": "application/x-www-form-urlencoded"}
+        )
+        request.set_formdata_body(data)
```

Comment:
Why not use request = HttpRequest(
            "POST", self._token_endpoint, headers={"Content-Type": "application/x-www-form-urlencoded"}, form_content=data
        )?

Reply:
I had the (mistaken) idea that if I used the `data` argument, I'd get the wrong content type.

Reply:
Depends on which kwarg you use. :)

We have code:

        if content is not None:
            if isinstance(content, ET.Element):
                request.set_xml_body(content)
            else:
                try:
                    request.set_json_body(content)
                except TypeError:
                    request.data = content

        if form_content:
            request.set_formdata_body(form_content)
        elif stream_content:
            request.set_streamed_data_body(stream_content)